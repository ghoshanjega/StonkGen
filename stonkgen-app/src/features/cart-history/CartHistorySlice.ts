import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import axios, { AxiosError } from "axios";
import { IBookCartItem, ICartItem } from "../../api-interface/Cart";
import { IApiError } from "../../api-interface/Error";
import { AppThunk, RootState } from "../../app/store";
import { ICartItemExtended } from "../cart/CartSlice";

export interface IGroupedCartItems  {
    sessionId : string;
    items : ICartItemExtended[]
} 

interface CartState {
  isLoading: boolean;
  error: AxiosError | null;
  cartItems: IGroupedCartItems[]  | null;
}

const initialState: CartState = {
  isLoading: true,
  error: null,
  cartItems: null,
};



export const cartHistorySlice = createSlice({
  name: "cartHistory",
  initialState,
  reducers: {
    fetchStarted: (state, action) => {
      state.isLoading = true;
    },
    fetchSuccessful: (state, action) => {
      state.isLoading = false;
      const cartItemsExtended = action.payload as IGroupedCartItems[];
      state.cartItems = cartItemsExtended;
    },
    fetchUnSuccessful: (state, action) => {
      state.isLoading = false;
      state.error = action.payload;
    },
  },
});

export const {
  fetchStarted,
  fetchSuccessful,
  fetchUnSuccessful,
} = cartHistorySlice.actions;


export const fetchCartItems = (params: any): AppThunk => (dispatch) => {
  dispatch(fetchStarted);
  return axios
    .get("http://localhost:3001/cart/history", {
      params,
    })
    .then(({ data }) => {
      dispatch(fetchSuccessful(data));
    })
    .catch((err: AxiosError) => {
      dispatch(fetchUnSuccessful(err));
    });
};


export const selectCartHistoryItems = (state: RootState) => state.cartHistory.cartItems;

export default cartHistorySlice.reducer;
