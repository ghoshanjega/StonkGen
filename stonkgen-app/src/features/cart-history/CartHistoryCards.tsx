import React from "react";
import { DisplayStatus } from "../cart/DisplayStatus";
import { IGroupedCartItems } from "./CartHistorySlice";

interface Props {
  cartItems: IGroupedCartItems[] | null;
}

export const CartHistoryCards = ({ cartItems }: Props) => {
  if (cartItems && cartItems.length > 0)
    return (
      <div>
        {cartItems?.map((session) => {
          var date = new Date(parseInt(session.sessionId));
          return (
            <div className="card mb-3" key={session.sessionId}>
              <div className="card-body">
                <p className="text-muted mb-2">{date.toUTCString()} </p>
                <table className="table table-sm">
                  {session.items.length > 0 && (
                    <thead>
                      <tr>
                        <th></th>
                        <th>Stock code</th>
                        <th>Execution side</th>
                        <th>Execution mode</th>
                        <th>Order price</th>
                        <th>amount</th>
                      </tr>
                    </thead>
                  )}
                  <tbody>
                    {session.items.map((item) => {
                      return (
                        <tr key={item.id}>
                          <td>
                            <DisplayStatus displayStatus={item.status} />
                          </td>
                          <td data-test="bloombergTickerLocal">{item.stock.bloombergTickerLocal}</td>
                          <td data-test="executionSide">{item.executionSide}</td>
                          <td data-test="executionMode">{item.executionModeBooked}</td>
                          <td data-test="priceBooked">{item.priceBooked + " " + item.stock.currency}</td>
                          <td data-test="amountBooked">{item.amountBooked}</td>
                        </tr>
                      );
                    })}
                  </tbody>
                </table>
              </div>
            </div>
          );
        })}
      </div>
    );
  else {
    return (
      <div className="jumbotron jumbotron-fluid">
        <div className="container">
          <p className="lead text-center" data-test="nothing">Nothing to show yet</p>
        </div>
      </div>
    );
  }
};
