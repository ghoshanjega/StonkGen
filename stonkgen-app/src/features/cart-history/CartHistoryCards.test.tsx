import React from "react";
import { shallow } from "enzyme";

import { CartHistoryCards } from "./CartHistoryCards";
import { IGroupedCartItems } from "./CartHistorySlice";

const mockItem: IGroupedCartItems[] = [{
  sessionId: "testSession",
  items: [
    {
      id: "1613486729355",
      executionSide: "sell",
      status: "booked",
      isLoading: false,
      amount: 1,
      displayPrice: 1,
      error: null,
      displayStatus: "prebooking",
      executionMode: "buy",
      selected: false,
      stock: {
        stockId: "7b93b1ab-5bc0-49d7-bd4f-3b1806e5750c",
        currency: "HKD",
        ric: "0900.HK",
        bloombergTicker: "900 HK",
        bloombergTickerLocal: "900 HK",
        name: "AEON CREDIT SERVICE (ASIA) CO",
        country: "Hong Kong",
        price: 338.74,
      },
      executionModeBooked: "limit",
      amountBooked: 6,
      priceBooked: 54,
    },
  ],
}];

const handleItemChange = () => {};

describe("Cart History Cards Component", () => {
  it("should render the cart Item", () => {
    let component = shallow(<CartHistoryCards cartItems={mockItem}  />);
    expect(component).toMatchSnapshot();
  });

  it("should match mock props", () => {
    let component = shallow(<CartHistoryCards cartItems={mockItem}  />);
    expect(component.find('[data-test="bloombergTickerLocal"]').props().children).toEqual(mockItem[0].items[0].stock.bloombergTickerLocal);
    expect(component.find('[data-test="executionSide"]').props().children).toEqual(mockItem[0].items[0].executionSide);
    expect(component.find('[data-test="priceBooked"]').props().children).toEqual(mockItem[0].items[0].priceBooked+" "+mockItem[0].items[0].stock.currency);
    expect(component.find('[data-test="amountBooked"]').props().children).toEqual(mockItem[0].items[0].amountBooked);
    expect(component.find('[data-test="executionMode"]').props().children).toEqual(mockItem[0].items[0].executionModeBooked);
  });

  it("should render null props", () => {
    let component = shallow(<CartHistoryCards cartItems={null}  />);
    expect(component.find('[data-test="nothing"]').props().children).toEqual("Nothing to show yet");
  });
});
