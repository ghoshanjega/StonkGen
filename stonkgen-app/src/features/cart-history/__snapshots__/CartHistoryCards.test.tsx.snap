// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart History Cards Component should render the cart Item 1`] = `ShallowWrapper {}`;

exports[`Cart Item Component should render the cart Item 1`] = `ShallowWrapper {}`;
